Keep heading during yaw convergence

bresch opened the pull request:
### Solved Problem
In altitude and position modes, when the yaw estimate is converging, the controller makes the drone yaw in order to follow the current setpoint. This is unintuitive for the pilot and it is preferable if the drone continues to fly towards the same physical direction.
[![Yaw correction (main)](https://img.youtube.com/vi/JODPqhQ_BVY/default.jpg)](https://youtu.be/JODPqhQ_BVY)


### Solution
Use an "unaided" yaw source based on corrected gyro integration to detect a quick change of the yaw estimate (using a high-pass filter) and compensate the yaw setpoint to make keep the correct physical heading.

In the video below, look at the heading indicator
[![No yaw correction (PR)](https://img.youtube.com/vi/Ml5zTLA_zfQ/default.jpg)](https://youtu.be/Ml5zTLA_zfQ)

In the plot below, see how the yaw setpoint is adjusted to maintain the current heading (starting from an error of 90 degrees)
![fast_convergence](https://github.com/PX4/PX4-Autopilot/assets/14822839/95d8a946-92ea-4b4c-9c92-4c46e2ac94b2)


### Changelog Entry
For release notes:
```
Improvement: Keep heading during yaw convergence
New parameter: -
Documentation: no
```

### Alternatives
Open the yaw attitude loop and rely on the yaw-rate tracking. I originally implemented this solution but I've seen many users having yaw spins on takeoff when the yaw loop isn't active.

### Test coverage
-sitl tests (also yaw emergency reset)

MaEtUgR commented on src/modules/flight_mode_manager/tasks/ManualAltitude/FlightTaskManualAltitude.cpp:
```diff
@@ -273,20 +275,62 @@ void FlightTaskManualAltitude::_ekfResetHandlerHeading(float delta_psi)
-	_stick_yaw.generateYawSetpoint(_yawspeed_setpoint, _yaw_setpoint, _sticks.getYawExpo(), _yaw,
-				       _is_yaw_good_for_control, _deltatime);
-	_acceleration_setpoint.xy() = _stick_tilt_xy.generateAccelerationSetpoints(_sticks.getPitchRoll(), _deltatime, _yaw,
+	float yaw_setpoint_uncorrected = wrap_pi(_yaw_setpoint - _yaw_correction);
+	const bool reset_required = _updateYawCorrection();
+
+	if (reset_required) {
+		_yaw_setpoint = NAN;
+		yaw_setpoint_uncorrected = NAN;
+	}
+
+	const float yaw_opposite_correction = wrap_pi(_yaw -
+					      _yaw_correction); // remove correction from the estimate in case it's used to generate the setpoint
+	_stick_yaw.generateYawSetpoint(_yawspeed_setpoint, yaw_setpoint_uncorrected, _sticks.getYawExpo(),
+				       yaw_opposite_correction,
+				       _deltatime);
+	_yaw_setpoint = wrap_pi(yaw_setpoint_uncorrected + _yaw_correction);
```
Maybe stupid question to ask but: In Stabilized, Altitude and Position mode the only thing we want to achieve is to lock yaw when no turn rate is commanded. How about we just control the unaided state which just has the goal to not drift but doesn't represent a real-world heading to north. Wouldn't that amke everything easier since you don't have to adjust the setpoint on the fly in the right way but rather just have a possibility to lock yaw and the controller does the right thing. Here's some pseudo-code for the idea:

Client old:
`if (no_yaw_rate && !isfinite(_yaw_setpoint)) _yaw_setpoint = _yaw;`
Client new:
`_yaw_setpoint_locked = no_yaw_rate;`

Controller new:
`if (_yaw_setpoint_locked) yaw_error = unaided_heading - _unaided_heading_when_lock_started;`

dagar commented:
Is there an acceptable solution that keeps it simple for the consumers?

Something like during periods of yaw estimate converging could we simply switch to publishing the unaided heading instead (reinitialized), and then do a single clean reset once ready?

bresch commented:
> Is there an acceptable solution that keeps it simple for the consumers?
> Something like during periods of yaw estimate converging could we simply switch to publishing the unaided heading instead (reinitialized), and then do a single clean reset once ready?

We discussed it during the dev call, but I'm still adding it here for visibility:
This yaw handling is quite specific to manual modes and shouldn't be part of the estimator itself. I will rather move it in the `StickYaw` library if possible to make sure it also gets applied when in stabilized mode.

bresch commented:
Flight test on an x500: https://logs.px4.io/plot_app?log=f70ba069-99c5-4ba0-ad4c-ec209633ec31

With an initial yaw error of ~20 degrees, we can see that the real heading of the vehicle stays constant while the yaw estimate is converging.
![Screenshot from 2023-08-23 10-52-58](https://github.com/PX4/PX4-Autopilot/assets/14822839/54d9a1ad-abf4-4f47-821f-103912f5d586)

bresch commented:
Cleaned-up the commit history and added comments on the algorithm.

Good to go on my side.

bresch commented on src/modules/flight_mode_manager/tasks/FlightTask/FlightTask.cpp:
```diff
@@ -112,6 +112,7 @@ void FlightTask::_evaluateVehicleLocalPosition()
 		// yaw
 		_yaw = _sub_vehicle_local_position.get().heading;
+		_unaided_yaw = _sub_vehicle_local_position.get().unaided_heading;
 		_is_yaw_good_for_control = _sub_vehicle_local_position.get().heading_good_for_control;
```
For now I decided to still keep it for auto modes, but eventually I would like to remove it completely

MaEtUgR reviewed (dismissed):
As we discussed the estimator side amendment of the published state would be quite a bit more involved. But it would still be nice to move the added functionality into the `StickYaw` class. Then it would be very clear what's the interface, the resets would happen internally and it would be easier to reuse for stabilized or other modes. If it should not apply for auto modes (yet) then we could just not call the additional correction function there or have a configuration of the class to disable. Could we still do that or move that to the next step?

bresch commented:
- Moved logic to `StickYaw`
- Moved computation of `_unaided_yaw` to ouptut predictor to be correctly time-aligned with the actual `yaw` estimate

@MaEtUgR Could you review again please?

MaEtUgR commented on src/modules/flight_mode_manager/tasks/Utility/StickYaw.hpp:
```diff
@@ -48,11 +48,22 @@ class StickYaw : public ModuleParams
 
 private:
 	AlphaFilter<float> _yawspeed_filter;
+	AlphaFilter<float> _yaw_error_lpf{0.01f};
+	bool _yaw_estimate_converging{false};
+	float _yaw_error_ref{0.f};
+	float _yaw_correction{0.f};
+
+	static constexpr float _kYawErrorTimeConstant{1.f};
+	static constexpr float _kYawErrorChangeThreshold{radians(1.f)};
```
Any parameter description that says more then the name would help 👀

MaEtUgR replied:
Was added ✅

MaEtUgR commented on src/modules/flight_mode_manager/tasks/Utility/StickYaw.cpp:
```diff
@@ -35,23 +35,80 @@
+
+	const bool was_converging = _yaw_estimate_converging;
+	_yaw_estimate_converging = fabsf(yaw_error_hpf) > _kYawErrorChangeThreshold;
+	_yaw_error_lpf.update(yaw_error);
```
This one could be moved two lines above after it was used no? Just a detail.

bresch replied:
Yes, good point; I'll move it

MaEtUgR reviewed (approved):
Not simple but should work 😃
Thanks for moving things into the StickYaw class 🙏

bresch commented on src/modules/ekf2/EKF/output_predictor.cpp:
```diff
@@ -237,6 +237,11 @@ void OutputPredictor::calculateOutputStates(const uint64_t time_us, const Vector
+
+	// update auxiliary yaw estimate
+	const Vector3f unbiased_delta_angle = delta_angle - delta_angle_bias_scaled;
+	const float spin_del_ang_D = unbiased_delta_angle.dot(Vector3f(_R_to_earth_now.row(2)));
```
Yes, basically you project the body frame angular velocity on the down axis of the local frame to get the yaw rate.

bresch commented on src/modules/flight_mode_manager/tasks/ManualAltitude/FlightTaskManualAltitude.cpp:
```diff
@@ -273,15 +274,20 @@ void FlightTaskManualAltitude::_ekfResetHandlerHeading(float delta_psi)
+				       _yaw,
+				       _deltatime, _unaided_yaw);
+
+	_acceleration_setpoint.xy() = _stick_tilt_xy.generateAccelerationSetpoints(_sticks.getPitchRoll(), _deltatime,
+				      _yaw,
 				      _yaw_setpoint);
```
Oh, that probably came from the first implementation followed by a make format, followed by a refactoring; will fix it.

bresch commented:
@MaEtUgR Thanks for the review, I addressed your comments.

MaEtUgR commented on src/modules/flight_mode_manager/tasks/ManualAltitude/FlightTaskManualAltitude.cpp:
```diff
@@ -273,15 +274,20 @@ void FlightTaskManualAltitude::_ekfResetHandlerHeading(float delta_psi)
-	_acceleration_setpoint.xy() = _stick_tilt_xy.generateAccelerationSetpoints(_sticks.getPitchRoll(), _deltatime, _yaw,
+	_stick_yaw.generateYawSetpoint(_yawspeed_setpoint, _yaw_setpoint, _sticks.getYawExpo(),
+				       _yaw,
+				       _deltatime, _unaided_yaw);
```
Corrected ✅
